staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -738,28 +747,56 @@ func (s *store) getStateFromObject(obj runtime.Object) (*objState, error) {
 	return state, nil
 }
 
-func (s *store) updateState(st *objState, userUpdate storage.UpdateFunc) (runtime.Object, uint64, error) {
-	ret, ttlPtr, err := userUpdate(st.obj, *st.meta)
+var errStaleTTL = errors.New("updateState needs current objState for TTL calculation")
+
+func (s *store) updateState(st *objState, userUpdate storage.UpdateFunc, mustCheckData bool) (runtime.Object, uint64, error) {
+	ret, ttlPtr, err := userUpdate(st.obj, st.meta)
 	if err != nil {
 		return nil, 0, err
 	}
 
-	if err := s.versioner.PrepareObjectForStorage(ret); err != nil {
-		return nil, 0, fmt.Errorf("PrepareObjectForStorage failed: %v", err)
-	}
 	var ttl uint64
 	if ttlPtr != nil {
+		// if userUpdate asserts a TTL and we are using a cached object, it means that we passed a zero TTL
```

Comment:
I'm finding all the TTL states and their semantics difficult to keep track of. I'm concerned that this code is going to be difficult to maintain. Should introduce additional structs with more explicitly named states?

Reply:
I will write it out on paper and see if I can come up with something cleaner.